rust/src/pop3/pop3.rs

```diff
@@ -0,0 +1,498 @@
+                                    unsafe {
+                                        AppLayerRequestProtocolTLSUpgrade(flow);
+                                    };
+                                }
+                                sawp_pop3::Keyword::RETR => {
+                                    // Don't hold onto the whole email body
+
+                                    // TODO: pass off to mime parser
+                                    response.data.clear();
+                                }
+                                _ => {}
+                            }
+                        }
+                        tx.response = Some(response);
+                    }
+                    start = rem;
+                }
+                Ok((rem, None)) => {
+                    // Not enough data. This parser doesn't give us a good indication
+                    // of how much data is missing so just ask for one more byte so
+                    // parse is called as soon as more data is received.
+
+                    let consumed = input.len() - rem.len();
+                    let needed = rem.len() + 1;
+                    return AppLayerResult::incomplete(consumed as u32, needed as u32);
+                }
+                Err(SawpError {
+                    kind: SawpErrorKind::Incomplete(sawp::error::Needed::Size(needed)),
+                }) => {
+                    let consumed = input.len() - start.len();
+                    let needed = start.len() + needed.get();
+                    return AppLayerResult::incomplete(consumed as u32, needed as u32);
+                }
+                Err(SawpError {
+                    kind: SawpErrorKind::Incomplete(sawp::error::Needed::Unknown),
+                }) => {
+                    let consumed = input.len() - start.len();
+                    let needed = start.len() + 1;
+                    return AppLayerResult::incomplete(consumed as u32, needed as u32);
+                }
+                Err(_) => {
+                    return AppLayerResult::err();
+                }
+            }
+        }
+
+        // All input was fully consumed.
+        return AppLayerResult::ok();
+    }
+
+    fn on_request_gap(&mut self, _size: u32) {
+        self.request_gap = true;
+    }
+
+    fn on_response_gap(&mut self, _size: u32) {
+        self.response_gap = true;
+    }
+}
+
+/// Reasonably need at least 5 bytes to determine
+/// 3 bytes for short commands like 'TOP' or response '+OK'
+/// and 2 bytes for the CRLF.
+static MIN_PROBE_LEN: u32 = 5;
+
+/// Probe for a command or response
+fn probe(input: &[u8]) -> AppProto {
+    match POP3_PARSER.probe(input, Direction::Unknown) {
+        Status::Recognized => unsafe { ALPROTO_POP3 },
+        Status::Incomplete => ALPROTO_UNKNOWN,
+        Status::Unrecognized => ALPROTO_FAILED,
+    }
+}
+
+// C exports.
+
+/// C entry point for a probing parser.
+unsafe extern "C" fn rs_pop3_probing_parser(
```

Comment:
nit: function name should be `SCPop3ProbingParser`

Reply:
actually, since it's not pub I think we then use a rust style, so it should just drop the `rs_` prefix

Reply:
same for the other private `rs_` funcs below